src/main/java/org/opensearch/security/configuration/DlsFlsFilterLeafReader.java

```diff
@@ -65,157 +58,73 @@
 
         this.shardId = shardId;
-        flsEnabled = includesExcludes != null && !includesExcludes.isEmpty();
-
-        if (flsEnabled) {
-
-            final FieldInfos infos = delegate.getFieldInfos();
-            this.includesSet = new HashSet<>(includesExcludes.size());
-            this.excludesSet = new HashSet<>(includesExcludes.size());
-
-            for (final String incExc : includesExcludes) {
-                if (canOptimize && (incExc.indexOf('.') > -1 || incExc.indexOf('*') > -1)) {
-                    canOptimize = false;
-                }
-
-                final char firstChar = incExc.charAt(0);
-
-                if (firstChar == '!' || firstChar == '~') {
-                    excludesSet.add(incExc.substring(1));
-                    excludesSet.add(incExc.substring(1) + KEYWORD);
-                } else {
-                    includesSet.add(incExc);
-                }
-            }
-
-            int i = 0;
-            final FieldInfo[] fa = new FieldInfo[infos.size()];
-
-            if (canOptimize) {
-                if (!excludesSet.isEmpty()) {
-                    for (final FieldInfo info : infos) {
-                        if (!excludesSet.contains(info.name)) {
-                            fa[i++] = info;
-                        }
-                    }
-                } else {
-                    for (final String inc : includesSet) {
-                        FieldInfo f;
-                        if ((f = infos.fieldInfo(inc)) != null) {
-                            fa[i++] = f;
-                        }
-                    }
-                }
-            } else {
-                if (!excludesSet.isEmpty()) {
-                    WildcardMatcher matcher = WildcardMatcher.from(excludesSet);
-                    for (final FieldInfo info : infos) {
-                        if (!matcher.test(info.name)) {
-                            fa[i++] = info;
-                        }
-                    }
+        this.flsRule = flsRule;
+        this.fmRule = fmRule;
+        this.metaFields = metaFields;
 
-                    this.excludes = excludesSet.toArray(EMPTY_STRING_ARRAY);
+        try {
+            if (!flsRule.isAllowAll()) {
+                FieldInfos originalFieldInfos = delegate.getFieldInfos();
+                List<FieldInfo> restrictedFieldInfos = new ArrayList<>(originalFieldInfos.size());
 
-                } else {
-                    WildcardMatcher matcher = WildcardMatcher.from(includesSet);
-                    for (final FieldInfo info : infos) {
-                        if (matcher.test(info.name)) {
-                            fa[i++] = info;
-                        }
+                for (FieldInfo fieldInfo : originalFieldInfos) {
+                    if (metaFields.contains(fieldInfo.name) || flsRule.isAllowed(fieldInfo.name)) {
+                        restrictedFieldInfos.add(fieldInfo);
```

Comment:
Yes, this will influence the values being read while searching. The APIs were looking at here a quite low level Lucene APIs - that's why they need to be considered as one big construct which achieves it's goals cooperatively. 

It's a bit difficult to name the purpose of each method individually.

Still, the `FieldInfo` usage is not new, but has just fundametally changed due to the more expressive `FlsRule` functionality. Before, it looked like this:

https://github.com/opensearch-project/security/blob/28000a3c58f15665307ae55dafa7351e3241436f/src/main/java/org/opensearch/security/configuration/DlsFlsFilterLeafReader.java#L162-L210

Reply:
so this loosely equates to includes rule? I'm just trying to understand what it would like look like for a request that comes in with a set of either include or exclude or both fields.

Reply:
No, the whole logic is replicated in the new code, both include/exclude scenarios are evaluated. However the include/exclude logic has been encapsulated here:

https://github.com/opensearch-project/security/blob/1fef11be731e5e2a3c06c5d565faf23525ffe0d6/src/main/java/org/opensearch/security/privileges/dlsfls/FieldPrivileges.java#L190-L215